forge/csrc/autograd/autograd.cpp

```diff
@@ -719,6 +721,161 @@ NodeContext autograd_engine::create_input(
     return NodeContext(node);
 }
 
+// From the node in the forward graph, create a recompute node for the backward graph.
+// This will also connect the newly created recompute node to its operands (based on the operands of the fwd node).
+// NOTE: it is expected that the create_recompute_op() is called in topological order.
+static Node *create_recompute_op(
+    graphlib::Graph *graph, Node *fwd_node, std::unordered_map<Node *, Node *> &fwd_to_recompute)
+{
+    graphlib::PyOpNode *recompute_node = graph->add_node(
```

Comment:
Just a thought here - maybe we can put these few lines of creating recompute node inside method create_recompute_node. I remember when we worked with buff queues we had few places in stack where we had similar code for creating buffering queues and it was chaotic. OFC recompute node is hopefully created only here so this shouldn't be the case...

Reply:
Yeah, I see your point with the buffering queues, but the recompute is not a special kind of node. And it is tightly coupled with autograd, so I don't think there should be any problems like with buffering queues.

However, I will modify this code to simply clone the forward node and change its epoch; instead of creating and adding the recompute node manually.